File: ui/web/src/data/passages.js
// Passages — span-relative intensity envelopes layered over the chapter run.
//
// JS mirror of `forge/passages.py` (KEEP IN SYNC). The shape math here only
// *previews* the curve in the UI; the actual scaling is baked at generation /
// export by the Python engine (e-stim volume + multi-axis amplitude). A passage
// answers "how much" over a scene-scale span; per-chapter Character/Mechanical
// answers "what". See project_passage_arcs_cross_modality memory.

// The five envelope shapes. Steady is the identity (no-op) default.
export const PASSAGE_SHAPES = [
  { id: 'steady',  label: 'Steady',  glyph: '▄▄▄▄',  desc: 'No change — the neutral default. A Steady passage does nothing.' },
  { id: 'build',   label: 'Build',   glyph: '▁▃▅▇',  desc: 'Rises from floor to ceiling across the span.' },
  { id: 'sustain', label: 'Sustain', glyph: '▇▇▇▇',  desc: 'Holds at the ceiling for the whole span.' },
  { id: 'release', label: 'Release', glyph: '▇▅▃▁',  desc: 'Eases from ceiling down to floor — the wind-down ("afterglow").' },
  { id: 'swell',   label: 'Swell',   glyph: '▁▅▇▅▁', desc: 'Rises to a peak at the middle, then eases back.' },
];

export const SHAPE_BY_ID = Object.fromEntries(PASSAGE_SHAPES.map((s) => [s.id, s]));

// The arc is ONE eased parametric envelope; the five named shapes are presets
// of it (rise_point, fall_point). Mirrors forge/passages.py::SHAPE_POINTS.
export const SHAPE_POINTS = {
  steady:  [0.0, 1.0],
  build:   [1.0, 1.0],
  sustain: [0.0, 1.0],
  release: [0.0, 0.0],
  swell:   [0.5, 0.5],
};

function smoothstep(t) {
  const x = t < 0 ? 0 : (t > 1 ? 1 : t);
  return x * x * (3 - 2 * x);
}

// Eased multiplier at `frac`: rise floor→ceiling by `risePoint`, hold to
// `fallPoint`, ease back to floor by the end. Mirrors envelope_factor().
export function envelopeFactor(frac, floor, ceiling, risePoint, fallPoint) {
  const f = frac < 0 ? 0 : (frac > 1 ? 1 : frac);
  const lo = floor;
  const hi = ceiling;
  let rp = risePoint < 0 ? 0 : (risePoint > 1 ? 1 : risePoint);
  let fp = fallPoint < 0 ? 0 : (fallPoint > 1 ? 1 : fallPoint);
  if (fp < rp) fp = rp;
  if (f <= rp) return rp <= 0 ? hi : lo + (hi - lo) * smoothstep(f / rp);
  if (f <= fp) return hi;
  if (fp >= 1) return hi;
  return hi - (hi - lo) * smoothstep((f - fp) / (1 - fp));
}

// Multiplier at fractional position `frac` (0..1). Explicit rise/fall win;
// else derived from the named shape. Mirrors forge/passages.py::shape_factor.
export function shapeFactor(shape, frac, floor, ceiling, risePoint, fallPoint) {
  const s = (shape || 'steady').toLowerCase();
  if (s === 'steady' && risePoint == null && fallPoint == null) return 1.0;
  const [rpDef, fpDef] = SHAPE_POINTS[s] || [1.0, 1.0];
  const rp = risePoint == null ? rpDef : risePoint;
  const fp = fallPoint == null ? fpDef : fallPoint;
  return envelopeFactor(frac, floor, ceiling, rp, fp);
}

// Effective rise/fall for a record (explicit, else the shape's preset points).
export function effectivePoints(rec) {
  const [rpDef, fpDef] = SHAPE_POINTS[(rec?.shape || 'steady').toLowerCase()] || [1.0, 1.0];
  return [rec?.risePoint ?? rpDef, rec?.fallPoint ?? fpDef];
}

// Sample a passage's envelope as N points in [0,1] (for sparkline previews).
export function sampleEnvelope(shape, floor, ceiling, n = 24, risePoint, fallPoint) {
  const out = [];
  for (let i = 0; i < n; i += 1) {
    const frac = n === 1 ? 0 : i / (n - 1);
    out.push(shapeFactor(shape, frac, floor, ceiling, risePoint, fallPoint));
  }
  return out;
}

let _uid = 0;
export function newPassageId() {
  _uid += 1;
  return `p${Date.now().toString(36)}_${_uid}`;
}

// A fresh passage spanning the given chapter index range, defaulting to Build.
// Carries explicit rise/fall handles (the parametric model) seeded from the
// shape so a hand-made passage is fully defined.
export function makePassage(beginIdx, endIdx, shape = 'build') {
  const [risePoint, fallPoint] = SHAPE_POINTS[shape] || [1.0, 1.0];
  return { id: newPassageId(), shape, beginIdx, endIdx, floor: 0.4, ceiling: 1.0, risePoint, fallPoint };
}

// ── Passage PRESETS — the resolved "one shared arc, preset-selected" model ──
// A passage is the depth/density of the SECONDARY world: ONE arc set across the
// whole chapter run that gives e-stim / mechanical / body their overall
// direction (each layer performs it its own way). Beta = presets only, same UI
// as Generate's Range/Pace lanes. Each preset maps onto the EXISTING passage
// data model (a single full-span passage of one shape), so forge/passages.py is
// untouched. See DESIGN_DECISIONS.md (Passages section).
// Each preset is a tuple of the four envelope params (start/floor, peak/ceiling,
// risePoint, fallPoint) over one full-span passage — "our own set", same spirit
// as Generate's Range/Pace presets. The sliders fine-tune any of them.
export const PASSAGE_PRESETS = [
  { id: 'hold',      label: 'Hold steady',     shape: 'steady',  floor: 1.0,  ceiling: 1.0,  risePoint: 0.0,  fallPoint: 1.0,  hint: 'no arc — flat throughout' },
  { id: 'hum',       label: 'Warm hum',        shape: 'sustain', floor: 0.55, ceiling: 0.58, risePoint: 0.0,  fallPoint: 1.0,  hint: 'a low, steady presence' },
  { id: 'build',     label: 'Build the charge', shape: 'build',  floor: 0.3,  ceiling: 1.0,  risePoint: 1.0,  fallPoint: 1.0,  hint: 'rises across the whole run' },
  { id: 'hold-high', label: 'Build & hold',    shape: 'build',   floor: 0.35, ceiling: 1.0,  risePoint: 0.55, fallPoint: 1.0,  hint: 'climbs, then holds at the top' },
  { id: 'edge',      label: 'Edge & release',  shape: 'swell',   floor: 0.35, ceiling: 1.0,  risePoint: 0.5,  fallPoint: 0.6,  hint: 'peak in the middle, then ease' },
  { id: 'release',   label: 'Long release',    shape: 'release', floor: 0.25, ceiling: 0.95, risePoint: 0.0,  fallPoint: 0.15, hint: 'starts high, eases down' },
];

// Build the passages array for a preset over `n` chapters. The neutral
// "Hold steady" is a no-op → no passage rows at all.
export function passagesForPreset(presetId, n) {
  const pr = PASSAGE_PRESETS.find((p) => p.id === presetId);
  if (!pr || pr.shape === 'steady' || n <= 0) return [];
  const p = makePassage(0, Math.max(0, n - 1), pr.shape);
  return [{ ...p, floor: pr.floor, ceiling: pr.ceiling, risePoint: pr.risePoint, fallPoint: pr.fallPoint }];
}

// Sample a preset's envelope (0..1) for the read-only lane preview.
export function presetSamples(presetId, count = 48) {
  const pr = PASSAGE_PRESETS.find((p) => p.id === presetId) || PASSAGE_PRESETS[0];
  return sampleEnvelope(pr.shape, pr.floor, pr.ceiling, count, pr.risePoint, pr.fallPoint);
}

// ── Resolve + sample at absolute time (mirrors forge/passages.py) ──
// These let the UI show the SAME multiplier the Python engine bakes, so the
// provenance badge / per-channel tag numbers match the rendered output.

// Resolve authored passage records to absolute-time spans against the chapter
// list. `chapters` carry atMs/endMs (camelCase) or at_ms/end_ms. Steady and
// degenerate spans are dropped (no-ops). Mirrors resolve_passages().
export function resolvePassageSpans(passages, chapters) {
  const n = chapters?.length || 0;
  if (!n) return [];
  const out = [];
  for (const rec of passages || []) {
    const shape = (rec.shape || 'steady').toLowerCase();
    if (shape === 'steady') continue;
    let bi = Number(rec.beginIdx ?? 0);
    let ei = Number(rec.endIdx ?? n - 1);
    if (Number.isNaN(bi) || Number.isNaN(ei)) continue;
    bi = Math.max(0, Math.min(bi, n - 1));
    ei = Math.max(0, Math.min(ei, n - 1));
    if (ei < bi) { const t = bi; bi = ei; ei = t; }
    const begin = chapters[bi];
    const end = chapters[ei];
    const lo = begin.atMs ?? begin.at_ms ?? begin.start_ms;
    const hi = end.endMs ?? end.end_ms;
    if (lo == null || hi == null || hi <= lo) continue;
    out.push({
      lo, hi, shape, beginIdx: bi, endIdx: ei,
      floor: Number(rec.floor ?? 0.2), ceiling: Number(rec.ceiling ?? 1.0),
      risePoint: rec.risePoint != null ? Number(rec.risePoint) : undefined,
      fallPoint: rec.fallPoint != null ? Number(rec.fallPoint) : undefined,
    });
  }
  return out;
}

// Combined multiplier at absolute time `t` (ms) over resolved spans. First
// covering span wins; no cover → 1.0. Mirrors factor_at().
export function passageFactorAt(spans, t) {
  for (const p of spans || []) {
    if (p.lo == null || p.hi == null || p.hi <= p.lo) continue;
    if (t >= p.lo && t <= p.hi) {
      const frac = (t - p.lo) / Math.max(1, p.hi - p.lo);
      return shapeFactor(p.shape, frac, p.floor, p.ceiling, p.risePoint, p.fallPoint);
    }
  }
  return 1.0;
}

// What passage (if any) shapes chapter `idx`, and by how much? Returns the
// covering span plus the multiplier at the chapter's MIDPOINT (representative)
// and at its start/end (the trajectory through the chapter), or null when no
// live passage covers it.
export function passageInfoForChapter(passages, chapters, idx) {
  const n = chapters?.length || 0;
  if (idx < 0 || idx >= n) return null;
  const spans = resolvePassageSpans(passages, chapters);
  if (!spans.length) return null;
  const ch = chapters[idx];
  const lo = ch.atMs ?? ch.at_ms ?? ch.start_ms;
  const hi = ch.endMs ?? ch.end_ms;
  if (lo == null || hi == null) return null;
  const mid = (lo + hi) / 2;
  const span = spans.find((s) => mid >= s.lo && mid <= s.hi);
  if (!span) return null;
  return {
    shape: span.shape,
    beginIdx: span.beginIdx, endIdx: span.endIdx,
    floor: span.floor, ceiling: span.ceiling,
    factor: passageFactorAt([span], mid),
    factorStart: passageFactorAt([span], lo),
    factorEnd: passageFactorAt([span], hi),
  };
}

// ── Single full-span "arc" ⇄ passages — the slider editor's model ──
// The arc lane edits ONE full-span passage via four params (start depth /
// peak / rise point / fall point). These convert to/from the passages array
// so presets and sliders share one representation.
const _ARC_EPS = 0.001;
const _clamp01 = (v) => (v < 0 ? 0 : (v > 1 ? 1 : v));

function _inferArcShape(floor, ceiling, rp, fp) {
  if (ceiling <= floor + _ARC_EPS) return 'sustain';       // flat line
  if (fp < 1 - _ARC_EPS) return rp > _ARC_EPS ? 'swell' : 'release';
  return rp < _ARC_EPS ? 'sustain' : 'build';
}

// Current arc params from the passages array (a single full-span passage), or
// a neutral flat-at-full arc when empty/multi-span.
export function arcFromPassages(passages) {
  if (passages && passages.length === 1) {
    const p = passages[0];
    const [rp, fp] = effectivePoints(p);
    return { floor: p.floor ?? 0.3, ceiling: p.ceiling ?? 1.0, risePoint: rp, fallPoint: fp };
  }
  return { floor: 1.0, ceiling: 1.0, risePoint: 0.0, fallPoint: 1.0 };
}

// Build a one-passage array from arc params over `n` chapters. A flat arc at
// full (ceiling≈floor≈1) is the neutral Hold → no passage. Enforces
// ceiling≥floor and fallPoint≥risePoint.
export function passagesFromArc(arc, n) {
  if (n <= 0) return [];
  const floor = _clamp01(arc.floor);
  const ceiling = Math.max(floor, _clamp01(arc.ceiling));
  if (ceiling >= 0.999 && floor >= 0.999) return [];
  const rp = _clamp01(arc.risePoint);
  const fp = Math.max(rp, _clamp01(arc.fallPoint));
  const shape = _inferArcShape(floor, ceiling, rp, fp);
  const p = makePassage(0, Math.max(0, n - 1), shape);
  return [{ ...p, floor, ceiling, risePoint: rp, fallPoint: fp }];
}

// The shape a passage reads as, inferred from its arc params (so the per-span
// editor can show a glyph without a shape dropdown).
export function arcShapeId(rec) {
  const floor = _clamp01(rec?.floor ?? 0.2);
  const ceiling = Math.max(floor, _clamp01(rec?.ceiling ?? 1.0));
  const [rp, fp] = effectivePoints(rec);
  if (ceiling >= 0.999 && floor >= 0.999) return 'steady';
  return _inferArcShape(floor, ceiling, _clamp01(rp), Math.max(_clamp01(rp), _clamp01(fp)));
}

// Which preset (if any) does the current passages array represent? Empty = the
// neutral "Hold steady"; a single full-span passage matching a preset's shape +
// floor/ceiling = that preset; anything hand-edited = null (no pill highlighted).
export function activePassagePreset(passages) {
  if (!passages || passages.length === 0) return 'hold';
  if (passages.length !== 1) return null;
  const p = passages[0];
  const [prp, pfp] = effectivePoints(p);
  const hit = PASSAGE_PRESETS.find((pr) => {
    const [rp, fp] = effectivePoints(pr);
    return pr.shape === p.shape
      && Math.abs(pr.floor - p.floor) < 0.001
      && Math.abs(pr.ceiling - p.ceiling) < 0.001
      && Math.abs(rp - prp) < 0.001
      && Math.abs(fp - pfp) < 0.001;
  });
  return hit ? hit.id : null;
}

File: forge/passages.py
"""Passages — span-relative intensity envelopes layered over the chapter run.

A *passage* covers a contiguous run of chapters and emits a multiplier in
``[floor, ceiling]`` that scales the intensity of whatever modality consumes it:
the e-stim **volume** channels and the multi-axis (mechanical) amplitude. The
envelope is **span-relative and length-independent** — a Build travels the full
floor->ceiling across the *passage*, not per chapter — so a slow scene-scale
contour survives even when chapters are only 3-5 min each (the per-chapter ramp
is imperceptible: Edger's volume ramp is 15%/hour ≈ 1% over a 4-min chapter,
which is exactly the gap passages fill).

Per-chapter assignment (Channels: Character/Mechanical) answers *what*; a passage
answers *how much*. They compose: the passage multiplier rides on top of the
generated channels just before device clamping. See the
``project_passage_arcs_cross_modality`` / ``project_funscriptforge_post_beta``
memory for the design.
"""

from __future__ import annotations

# The five envelope shapes. Each maps a fractional position ``frac`` in [0, 1]
# across the passage span to a multiplier in [floor, ceiling]. Steady is the
# identity (1.0 everywhere) so an unassigned/default passage is a no-op.
PASSAGE_SHAPES = ("steady", "build", "sustain", "release", "swell")

# e-stim channels that carry *intensity* (scale these); phase/position channels
# (alpha/beta/frequency/...) pass through untouched. Mirrors
# ``stim_config._VOLUME_CHANNELS``.
ESTIM_INTENSITY_CHANNELS = ("volume", "volume-prostate")

# Neutral position for a multi-axis (rotary/linear) channel — amplitude is the
# deviation from this. Scaling toward it reduces motion without shifting center.
AXIS_CENTER = 50


# The arc is ONE parametric envelope: rise from ``floor`` to ``ceiling``,
# finishing at ``rise_point``; hold at ``ceiling`` until ``fall_point``; ease
# back to ``floor`` by the end. The five named shapes are just presets of it
# (so old ``<stem>.passages.json`` records without rise/fall keep working):
#   build   rise=1.0 fall=1.0  (climbs the whole span, never falls)
#   release rise=0.0 fall=0.0  (starts at the peak, eases the whole span)
#   sustain rise=0.0 fall=1.0  (instantly at the peak, holds)
#   swell   rise=0.5 fall=0.5  (a single rounded hump in the middle)
#   steady  identity (1.0) — the no-op default
SHAPE_POINTS = {
    "steady":  (0.0, 1.0),
    "build":   (1.0, 1.0),
    "sustain": (0.0, 1.0),
    "release": (0.0, 0.0),
    "swell":   (0.5, 0.5),
}


def _smoothstep(t: float) -> float:
    """Cubic ease 3t²-2t³ — rounded shoulders, no hard corners. Clamped 0..1."""
    t = 0.0 if t < 0.0 else (1.0 if t > 1.0 else t)
    return t * t * (3.0 - 2.0 * t)


def envelope_factor(frac: float, floor: float, ceiling: float,
                    rise_point: float, fall_point: float) -> float:
    """Eased multiplier at fractional position ``frac`` (0..1) along the span.

    Rise ``floor``->``ceiling`` over [0, rise_point], hold at ``ceiling`` over
    [rise_point, fall_point], ease back to ``floor`` over [fall_point, 1].
    Both transitions use a smoothstep so the baked arc is gently curved, not a
    straight ramp / pointy triangle.
    """
    f = 0.0 if frac < 0.0 else (1.0 if frac > 1.0 else frac)
    lo = float(floor)
    hi = float(ceiling)
    rp = 0.0 if rise_point < 0.0 else (1.0 if rise_point > 1.0 else rise_point)
    fp = 0.0 if fall_point < 0.0 else (1.0 if fall_point > 1.0 else fall_point)
    if fp < rp:
        fp = rp
    if f <= rp:
        return hi if rp <= 0.0 else lo + (hi - lo) * _smoothstep(f / rp)
    if f <= fp:
        return hi
    if fp >= 1.0:
        return hi
    return hi - (hi - lo) * _smoothstep((f - fp) / (1.0 - fp))


def shape_factor(shape: str, frac: float, floor: float, ceiling: float,
                 rise_point=None, fall_point=None) -> float:
    """Multiplier at ``frac`` for a passage. Explicit ``rise_point``/
    ``fall_point`` win; otherwise they derive from the named ``shape`` (so the
    five shapes are presets of the one eased envelope). ``steady`` is the
    identity (1.0), the no-op default."""
    s = (shape or "steady").lower()
    if s == "steady" and rise_point is None and fall_point is None:
        return 1.0
    rp_def, fp_def = SHAPE_POINTS.get(s, (1.0, 1.0))
    rp = rp_def if rise_point is None else float(rise_point)
    fp = fp_def if fall_point is None else float(fall_point)
    return envelope_factor(frac, floor, ceiling, rp, fp)


def factor_at(passages, t: int) -> float:
    """Combined passage multiplier at absolute time ``t`` (ms).

    Passages are treated as non-overlapping spans; the first one containing
    ``t`` wins. No covering passage -> 1.0 (untouched).
    """
    for p in passages or ():
        lo = p["lo"]
        hi = p["hi"]
        if lo is None or hi is None or hi <= lo:
            continue
        if lo <= t <= hi:
            frac = (t - lo) / max(1, hi - lo)
            return shape_factor(p.get("shape"), frac,
                                p.get("floor", 0.2), p.get("ceiling", 1.0),
                                p.get("rise_point"), p.get("fall_point"))
    return 1.0


def scale_estim(channel_name: str, actions: list[dict], passages) -> list[dict]:
    """Scale an e-stim channel's actions by the passage envelope.

    Only *intensity* (volume) channels are scaled — toward 0 (off). Phase and
    position channels pass through unchanged. No passages -> unchanged.
    """
    if not passages or channel_name not in ESTIM_INTENSITY_CHANNELS:
        return actions
    out = []
    for a in actions:
        factor = factor_at(passages, a["at"])
        out.append({"at": a["at"], "pos": int(round(a["pos"] * factor))})
    return out


def scale_axis(actions: list[dict], passages, center: int = AXIS_CENTER) -> list[dict]:
    """Scale a multi-axis (mechanical) channel's amplitude toward ``center``.

    Reduces motion amplitude by the passage envelope without shifting the
    neutral position (so a Release winds the geometry down to stillness, not
    to an off-center pose). No passages -> unchanged.
    """
    if not passages:
        return actions
    out = []
    for a in actions:
        factor = factor_at(passages, a["at"])
        pos = center + (a["pos"] - center) * factor
        out.append({"at": a["at"], "pos": int(round(pos))})
    return out


def resolve_passages(passage_records, chapters) -> list[dict]:
    """Map authored passage records to absolute-time spans for generation.

    ``passage_records`` is the list from ``<stem>.passages.json``; each carries
    a ``shape``, ``floor``/``ceiling`` (0..1) and a chapter span as 0-based
    ``beginIdx``/``endIdx`` (inclusive). ``chapters`` is the chapters-sidecar
    list (each with ``at_ms``/``end_ms``). Returns
    ``[{lo, hi, shape, floor, ceiling}]`` with ``steady`` / degenerate spans
    dropped (they are no-ops).
    """
    out = []
    n = len(chapters)
    if n == 0:
        return out
    for rec in passage_records or ():
        shape = (rec.get("shape") or "steady").lower()
        if shape == "steady":
            continue
        bi = rec.get("beginIdx", 0)
        ei = rec.get("endIdx", n - 1)
        try:
            bi = max(0, min(int(bi), n - 1))
            ei = max(0, min(int(ei), n - 1))
        except (TypeError, ValueError):
            continue
        if ei < bi:
            bi, ei = ei, bi
        begin = chapters[bi]
        end = chapters[ei]
        lo = begin.get("at_ms", begin.get("atMs", begin.get("start_ms")))
        hi = end.get("end_ms", end.get("endMs"))
        if lo is None or hi is None or hi <= lo:
            continue
        span = {
            "lo": int(lo),
            "hi": int(hi),
            "shape": shape,
            "floor": float(rec.get("floor", 0.2)),
            "ceiling": float(rec.get("ceiling", 1.0)),
        }
        # Optional explicit rise/fall handles (the parametric model); absent =
        # derive from the named shape inside shape_factor.
        if rec.get("risePoint") is not None:
            span["rise_point"] = float(rec["risePoint"])
        if rec.get("fallPoint") is not None:
            span["fall_point"] = float(rec["fallPoint"])
        out.append(span)
    return out
